fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/RecordMetaDataBuilder.java

```diff
@@ -642,7 +681,28 @@ private static boolean unionHasMessageType(@Nonnull Descriptors.Descriptor union
         return unionDescriptor.getFields().stream().anyMatch(field -> descriptor == field.getMessageType());
     }
 
-    private void fillUnionFields(Descriptors.Descriptor union, boolean processExtensionOptions) {
+    private void updateUnionFields(@Nonnull Descriptors.Descriptor union, boolean processExtensionOptions) {
+        for (Descriptors.FieldDescriptor unionField : union.getFields()) {
+            Descriptors.Descriptor descriptor = unionField.getMessageType();
+            if (recordTypes.containsKey(descriptor.getName())) {
+                RecordTypeBuilder oldRecordType = recordTypes.get(descriptor.getName());
+                RecordTypeBuilder newRecordType = new RecordTypeBuilder(descriptor, oldRecordType);
+                recordTypes.put(newRecordType.getName(), newRecordType); // update the record type builder
+                unionFields.remove(oldRecordType.getDescriptor());
+                unionFields.put(descriptor, unionField);
+            } else {
+                if (unionFields.containsKey(descriptor)) {
+                    // Union field was seen before but the record type is unknown? This must not happen.
+                    throw new MetaDataException("Unknown record type for union field " + unionField.getName());
+                }
+                // New field / record type.
+                processRecordType(unionField, processExtensionOptions);
+                unionFields.put(descriptor, unionField);
```

Comment:
> > Doesn't this need to set the since version of the record type it creates
> 
> `processRecordType` does it, right?

If I'm reading `processRecordType` correctly, it only does so if the `since_version` is included in an extension option of the record type, but that extension option should be considered deprecated. (Hm, maybe it's worth validating that the since version in the option is the same as the since version we're setting if it's defined?) In fact, I think this is effectively supposed to replace setting the `since_version` of a record type as the easiest way for users to make sure that that field is set properly.

> > (and probably bump the meta-data version)
> 
> Good point. `RecordMetaDataBuilder` never bumped the meta-data version when a new record type is added. But may be we should for `updateRecords`. Bumping it is definitely safe, I guess I don't know if there's any problem with **not** bumping it though. If the record types have no indexes, what would bumping it give us?

If the record type doesn't have any indexes, bumping the version doesn't give us much (and neither does setting the `since_version`, I suppose). But after updating the records, the user very well might add some new indexes on the new record type. (We can't know that here either way.) And if they do, making sure the `since_version` is set is necessary to to avoid triggering unnecessary index builds on the indexes on the new types.

In theory, we only need to bump it once for all records. (It's fine if all of the records have the same `since_version`.) But for the `since_version` to be valid, it should be strictly greater than the meta-data version prior to the meta-data  version before the records were updated. (Otherwise, it will claim that it was added to the meta-data prior to the version in which it was actually added.)

It's also good practice to bump the version whenever the meta-data is changed so that those changes can be updated, so maybe bumping it once in `updateRecords` is sufficient.